### PR TITLE
Delete cookie with invalid CSRF token

### PR DESCRIPTION
**Pull request by garas**
Alternate to #14492.

Delete cookie with invalid CSRF token to prevent user unable to post any forms persistently.

### CODE REVIEW

File: src/Http/Middleware/CsrfProtectionMiddleware.php [hunk 1]
```diff
@@ -193,6 +193,26 @@ public function createToken(): string
...
 
+    /**
+     * Verify CSRF token.
+     *
+     * @param string $token The CSRF token.
+     * @return bool
+     */
+    protected function _verifyToken(string $token): bool
+    {
+        if (strlen($token) <= self::TOKEN_VALUE_LENGTH) {
```
Comment (dereuromark) [inline src/Http/Middleware/CsrfProtectionMiddleware.php:204]: should those be mb_ operations? the token input could be any user provided input right?

**Review by markstory — COMMENTED** [review]
Nice work. Deleting the invalid cookies should work better as each user should only see one invalid csrf token error.

File: src/Http/Middleware/CsrfProtectionMiddleware.php [hunk 2]
```diff
@@ -193,6 +193,26 @@ public function createToken(): string
         return $value . hash_hmac('sha1', $value, Security::getSalt());
     }
 
+    /**
+     * Verify CSRF token.
```
Comment (othercorey) [inline src/Http/Middleware/CsrfProtectionMiddleware.php:197]: Can we add a little more explanation on the verification requirements? We want to document the hardened token change a little more than just a couple lines of code changing.

File: tests/TestCase/Http/Middleware/CsrfProtectionMiddlewareTest.php [hunk 3]
```diff
@@ -204,8 +205,19 @@ public function testInvalidTokenInHeader($method)
...
+            $middleware->process($request, $this->_getRequestHandler());
+
+            $this->fail();
+        } catch (InvalidCsrfTokenException $exception) {
+            $responseHeaders = $exception->responseHeader();
+
+            $this->assertArrayHasKey('Set-Cookie', $responseHeaders);
+
+            $cookie = Cookie::createFromHeaderString($responseHeaders['Set-Cookie']);
+            $this->assertSame('csrfToken', $cookie->getName(), 'Should delete cookie with invalid CSRF token');
```
Comment (othercorey) [inline tests/TestCase/Http/Middleware/CsrfProtectionMiddlewareTest.php:218]:
```suggestion
            $this->assertSame('csrfToken', $cookie->getName(), 'Should automatically delete cookie with invalid CSRF token');
```

File: src/Http/Middleware/CsrfProtectionMiddleware.php [hunk 4]
```diff
@@ -238,60 +251,51 @@ protected function _validateToken(ServerRequestInterface $request): void
...
      *
-     * @param mixed $post The request token.
-     * @param mixed $cookie The cookie token.
-     * @return bool
+     * @param string $value Cookie value
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request object.
+     * @return \Cake\Http\Cookie\Cookie
      */
-    protected function _compareToken($post, $cookie): bool
+    protected function _createCookie(string $value, ServerRequestInterface $request): Cookie
```
Comment (ADmad) [inline src/Http/Middleware/CsrfProtectionMiddleware.php:286]: Use `CookieInterface` instead for typehint?

**Comment by markstory** [comment]
Tested these changes out locally with CSRF tokens made in 4.0.5. The user will see one invalid CSRF token error and then get a new hardened token. Great work @garas :clap:

**Pull request merged**